### PR TITLE
rocAL - External Source Reader

### PR DESCRIPTION
**Pull request by SundarRajan98**
Added 3 Modes for external source reader whose sources varies accordingly :
• Mode 0 - File Names
• Mode 1 - Raw Buffer
• Mode 2 - Decoded Buffer
• Added C++ and python examples for all 3 modes

### CODE REVIEW

File: rocAL/include/api/rocal_api_data_loaders.h [hunk 1]
```diff
@@ -800,5 +800,30 @@ extern "C" RocalTensor ROCAL_API_CALL rocalJpegCaffe2LMDBRecordSourcePartialSing
                                                                                         bool loop = false,
                                                                                         RocalImageSizeEvaluationPolicy decode_size_policy = ROCAL_USE_MOST_FREQUENT_SIZE,
                                                                                         unsigned max_width = 0, unsigned max_height = 0);
+/*! \brief Creates JPEG external source image reader.
```
Comment (sampath-amd) [inline rocAL/include/api/rocal_api_data_loaders.h:803]: Is external source reader supported only for images?

File: rocAL/include/api/rocal_api_data_loaders.h [hunk 2]
```diff
@@ -800,5 +800,30 @@ extern "C" RocalTensor ROCAL_API_CALL rocalJpegCaffe2LMDBRecordSourcePartialSing
...
+extern "C" RocalTensor ROCAL_API_CALL rocalJpegExternalFileSource(RocalContext p_context,
+                                                                  const char* source_path,
+                                                                  RocalImageColor rocal_color_format,
+                                                                  bool is_output = false,
+                                                                  bool shuffle = false,
+                                                                  bool loop = false,
+                                                                  RocalImageSizeEvaluationPolicy decode_size_policy = ROCAL_USE_MOST_FREQUENT_SIZE,
+                                                                  unsigned max_width = 0, unsigned max_height = 0,
+                                                                  RocalDecoderType rocal_decoder_type = RocalDecoderType::ROCAL_DECODER_TJPEG,
+                                                                  RocalExtSourceMode external_source_mode = RocalExtSourceMode::ROCAL_EXTSOURCE_FNAME);
```
Comment (sampath-amd) [inline rocAL/include/api/rocal_api_data_loaders.h]: Please check if it is better to rename the enum class to RocalExternalSourceMode  instead of  RocalExtSourceMode

File: rocAL/include/api/rocal_api_types.h [hunk 3]
```diff
@@ -357,4 +357,11 @@ enum class RocalROICordsType {
 typedef struct {
     unsigned x1, y1, x2, y2;
 } RocalROICords;
+
+enum RocalExtSourceMode {
```
Comment (sampath-amd) [inline rocAL/include/api/rocal_api_types.h]: Add the Doxgen support for this new enum

File: rocAL/include/loaders/image/cifar10_data_loader.h [hunk 4]
```diff
@@ -47,6 +47,9 @@ class CIFAR10DataLoader : public LoaderModule {
     void shut_down() override;
     std::vector<std::vector<float>> &get_batch_random_bbox_crop_coords();
     void set_batch_random_bbox_crop_coords(std::vector<std::vector<float>> batch_crop_coords);
+    void feed_external_input(std::vector<std::string> input_images_names, std::vector<int> labels, std::vector<unsigned char *> input_buffer,
+                             std::vector<unsigned> roi_width, std::vector<unsigned> roi_height,
+                             unsigned int max_width, unsigned int max_height, int channels, ExternalFileMode mode, bool eos) override { return; }
```
Comment (sampath-amd) [inline rocAL/include/loaders/image/cifar10_data_loader.h]: The function return type is void. Why there is a return statement to this function?

File: rocAL/include/meta_data/caffe2_meta_data_reader.h [hunk 5]
```diff
@@ -43,6 +43,7 @@ class Caffe2MetaDataReader : public MetaDataReader {
     void print_map_contents();
     bool set_timestamp_mode() override { return false; }
     std::map<std::string, std::shared_ptr<MetaData>>& get_map_content() override { return (_map_content); }
+    void add_labels(std::vector<std::string> image_name, std::vector<int> label) override{};
```
Comment (sampath-amd) [inline rocAL/include/meta_data/caffe2_meta_data_reader.h]: please change variable names to 
- image_names
- labels

since it is a vector

File: rocAL/include/meta_data/external_source_label_reader.h [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+*/
+
+#pragma once
+#include <dirent.h>
+
+#include <map>
+
```
Comment (sampath-amd) [inline rocAL/include/meta_data/external_source_label_reader.h]: Please remove the empty blank lines

File: rocAL/include/pipeline/master_graph.h [hunk 7]
```diff
@@ -203,6 +207,11 @@ class MasterGraph {
     bool _offset;                                                                 // Returns normalized offsets ((encoded_bboxes*scale - anchors*scale) - mean) / stds in EncodedBBoxes that use std and the mean and scale arguments if offset="True"
     std::vector<float> _means, _stds;                                             //_means:  [x y w h] mean values for normalization _stds: [x y w h] standard deviations for offset normalization.
     bool _augmentation_metanode = false;
+    bool _external_source_eos = false;     // If last batch, _external_source_eos will true
+    bool _external_source_reader = false;  // Set to true if external source reader on
+    float _high_threshold = 0.5;
+    float _low_threshold = 0.4;
+    bool _allow_low_quality_matches = true;
```
Comment (sampath-amd) [inline rocAL/include/pipeline/master_graph.h]: why these changes are there in this PR?
float _high_threshold = 0.5;
float _low_threshold = 0.4;
bool _allow_low_quality_matches = true;

These are for box iou matcher if i am not wrong. Please remove these changes if not applicable

File: rocAL/include/readers/image/caffe2_lmdb_record_reader.h [hunk 8]
```diff
@@ -68,6 +68,12 @@ class Caffe2LMDBRecordReader : public Reader {
 
     Caffe2LMDBRecordReader();
 
+    //! return feed_data: not implemented
+    void feed_file_names(const std::vector<std::string>& file_names, size_t num_images, bool eos = false) override { return; }
```
Comment (sampath-amd) [inline rocAL/include/readers/image/caffe2_lmdb_record_reader.h]: The comment for the function is ambigious
Please clarify

File: rocAL/source/api/rocal_api_data_transfer.cpp [hunk 9]
```diff
@@ -79,6 +79,35 @@ void
...
+    unsigned int max_width,
+    unsigned int max_height,
+    int channels,
+    RocalExtSourceMode mode,
+    RocalTensorLayout layout,
+    bool eos) {
+    auto context = static_cast<Context*>(p_context);
+    try {
+        ExternalFileMode external_file_mode = (ExternalFileMode)mode;
+        RocalTensorlayout format = (RocalTensorlayout)layout;
```
Comment (sampath-amd) [inline rocAL/source/api/rocal_api_data_transfer.cpp]: Please use a static_cast instead of c style casting for L98, L99

File: rocAL/source/loaders/image/image_read_and_decode.cpp [hunk 10]
```diff
@@ -22,10 +22,13 @@ THE SOFTWARE.
 
 #include "image_read_and_decode.h"
 
+#include <unistd.h>
+
```
Comment (sampath-amd) [inline rocAL/source/loaders/image/image_read_and_decode.cpp]: why this include is required?

File: rocAL/source/loaders/image/image_read_and_decode.cpp [hunk 11]
```diff
@@ -92,6 +95,26 @@ void ImageReadAndDecode::create(ReaderConfig reader_config, DecoderConfig decode
...
+}
+
+void ImageReadAndDecode::feed_external_input(std::vector<std::string> input_images_names, std::vector<int> labels, std::vector<unsigned char *> input_buffer,
+                                             std::vector<unsigned> roi_width, std::vector<unsigned> roi_height,
+                                             unsigned int max_width, unsigned int max_height, int channels, ExternalFileMode mode, bool eos) {
+    std::vector<size_t> image_size;
+    image_size.reserve(roi_height.size());
+    for (unsigned int i = 0; i < roi_height.size(); i++) {
+        if (mode == ExternalFileMode::RAWDATA_UNCOMPRESSED)
+            image_size[i] = (max_width * max_height * channels);
```
Comment (sampath-amd) [inline rocAL/source/loaders/image/image_read_and_decode.cpp]: max_width * max_height * channels is a constant
Please pre compute this before for loop and use that variable here

File: rocAL/source/loaders/image/image_read_and_decode.cpp [hunk 12]
```diff
@@ -92,6 +95,26 @@ void ImageReadAndDecode::create(ReaderConfig reader_config, DecoderConfig decode
...
+        else if (mode == ExternalFileMode::RAWDATA_COMPRESSED)
+            image_size[i] = roi_height[i];
+    }
+    if (mode == ExternalFileMode::FILENAME)
+        _reader->feed_file_names(input_images_names, input_images_names.size(), eos);
+    else if (mode == ExternalFileMode::RAWDATA_COMPRESSED)
+        _reader->feed_data(input_buffer, image_size, mode, eos, {}, {}, max_width, max_height, channels);
```
Comment (sampath-amd) [inline rocAL/source/loaders/image/image_read_and_decode.cpp]: the values for roi_width, roi_height passed as {}?

File: rocAL/source/loaders/image/image_read_and_decode.cpp [hunk 13]
```diff
@@ -108,7 +131,7 @@ void ImageReadAndDecode::set_random_bbox_data_reader(std::shared_ptr<RandomBBoxC
     _randombboxcrop_meta_data_reader = randombboxcrop_meta_data_reader;
 }
 
-std::vector<std::vector<float>>&
+std::vector<std::vector<float>> &
```
Comment (sampath-amd) [inline rocAL/source/loaders/image/image_read_and_decode.cpp]: Please revert this change

File: rocAL/source/meta_data/external_source_label_reader.cpp [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+
+#include <string.h>
+
+#include <algorithm>
+#include <iostream>
+#include <utility>
+
+#include "commons.h"
+#include "exception.h"
+
```
Comment (sampath-amd) [inline rocAL/source/meta_data/external_source_label_reader.cpp]: Please remove the blank lines between includes
Applicable to all such instances

File: rocAL_pybind/amd/rocal/plugin/generic.py [hunk 15]
```diff
@@ -99,6 +99,37 @@ def __next__(self):
...
+                        "source_input_images": next(self.loader._external_source)[0],
+                        "labels": next(self.loader._external_source)[1],
+                        "input_batch_buffer": [],
+                        "roi_width": [],
+                        "roi_height": [],
+                        "decoded_width": self.loader._external_source_user_given_width,
+                        "decoded_height": self.loader._external_source_user_given_height,
+                        "channels": self.p,
+                        "external_source_mode": self.loader._external_source_mode,
+                        "rocal_tensor_layout": types.NCHW,
```
Comment (sampath-amd) [inline rocAL_pybind/amd/rocal/plugin/generic.py]: Why this is hardcoded?

File: rocAL_pybind/amd/rocal/types.py [hunk 16]
```diff
@@ -142,6 +147,10 @@
     SCALING_MODE_NOT_SMALLER: ("SCALING_MODE_NOT_SMALLER", SCALING_MODE_NOT_SMALLER),
     SCALING_MODE_NOT_LARGER: ("SCALING_MODE_NOT_LARGER", SCALING_MODE_NOT_LARGER),
 
+    EXTSOURCE_FNAME: ("EXTSOURCE_FNAME", EXTSOURCE_FNAME),
+    EXTSOURCE_RAW_COMPRESSED: ("EXTSOURCE_RAW_COMPRESSED", EXTSOURCE_RAW_COMPRESSED),
+    EXTSOURCE_RAW_UNCOMPRESSED: ("EXTSOURCE_RAW_UNCOMPRESSED", EXTSOURCE_RAW_UNCOMPRESSED),
+
```
Comment (sampath-amd) [inline rocAL_pybind/amd/rocal/types.py]: Please remove empty blank line

**Review by sampath-amd — COMMENTED** [review]
Please address review comments

File: rocAL_pybind/amd/rocal/types.py (hunk 16 above)
Comment (fgladwin) [reply rocAL_pybind/amd/rocal/types.py]: Done

File: rocAL/source/api/rocal_api_data_transfer.cpp (hunk 9 above)
Comment (fgladwin) [reply rocAL/source/api/rocal_api_data_transfer.cpp]: Done

File: rocAL/include/meta_data/external_source_label_reader.h (hunk 6 above)
Comment (fgladwin) [reply rocAL/include/meta_data/external_source_label_reader.h]: Done

File: rocAL/include/api/rocal_api_types.h (hunk 3 above)
Comment (fgladwin) [reply rocAL/include/api/rocal_api_types.h]: Done

File: rocAL/include/api/rocal_api_data_loaders.h (hunk 2 above)
Comment (fgladwin) [reply rocAL/include/api/rocal_api_data_loaders.h]: Done

File: rocAL/include/loaders/image/cifar10_data_loader.h (hunk 4 above)
Comment (fgladwin) [reply rocAL/include/loaders/image/cifar10_data_loader.h]: We are not returning anything in this function. Since this function doesn't have any specific implementation, return; is used as a placeholder.

File: rocAL/include/meta_data/caffe2_meta_data_reader.h (hunk 5 above)
Comment (fgladwin) [reply rocAL/include/meta_data/caffe2_meta_data_reader.h]: Done

File: rocAL/source/loaders/image/image_read_and_decode.cpp (hunk 10 above)
Comment (fgladwin) [reply rocAL/source/loaders/image/image_read_and_decode.cpp]: Removed the header

File: rocAL/source/loaders/image/image_read_and_decode.cpp (hunk 13 above)
Comment (fgladwin) [reply rocAL/source/loaders/image/image_read_and_decode.cpp]: Done

File: rocAL/source/meta_data/external_source_label_reader.cpp (hunk 14 above)
Comment (fgladwin) [reply rocAL/source/meta_data/external_source_label_reader.cpp]: Done

File: rocAL/source/loaders/image/image_read_and_decode.cpp (hunk 11 above)
Comment (fgladwin) [reply rocAL/source/loaders/image/image_read_and_decode.cpp]: Done

File: rocAL/include/api/rocal_api_data_loaders.h (hunk 1 above)
Comment (fgladwin) [reply rocAL/include/api/rocal_api_data_loaders.h:803]: currently it is supported only for images

File: rocAL/include/pipeline/master_graph.h (hunk 7 above)
Comment (fgladwin) [reply rocAL/include/pipeline/master_graph.h]: Done

File: rocAL/source/loaders/image/image_read_and_decode.cpp (hunk 12 above)
Comment (fgladwin) [reply rocAL/source/loaders/image/image_read_and_decode.cpp]: For mode 1 the roi_width and roi_height is empty

File: rocAL_pybind/amd/rocal/plugin/generic.py (hunk 15 above)
Comment (fgladwin) [reply rocAL_pybind/amd/rocal/plugin/generic.py]: @swetha097 can you please check this comment

File: rocAL/include/loaders/image/image_read_and_decode.h [hunk 17]
```diff
@@ -98,4 +101,5 @@ class ImageReadAndDecode {
     std::shared_ptr<RandomBBoxCrop_MetaDataReader> _randombboxcrop_meta_data_reader = nullptr;
     pCropCord _CropCord;
     RocalRandomCropDecParam *_random_crop_dec_param = nullptr;
+    StorageType _storage_type;
```
Comment (shobiloganat) [inline rocAL/include/loaders/image/image_read_and_decode.h]: Check the naming convention here.

File: rocAL/include/readers/image/caffe2_lmdb_record_reader.h (hunk 8 above)
Comment (shobiloganat) [reply rocAL/include/readers/image/caffe2_lmdb_record_reader.h]: remove the comment

File: rocAL/include/readers/image/caffe2_lmdb_record_reader.h [hunk 18]
```diff
@@ -68,6 +68,12 @@ class Caffe2LMDBRecordReader : public Reader {
...
+
+    //! return feed_raw_data: not implemented
```
Comment (shobiloganat) [inline rocAL/include/readers/image/caffe2_lmdb_record_reader.h]: Please add comments to this API.

File: rocAL/include/readers/image/caffe_lmdb_record_reader.h [hunk 19]
```diff
@@ -68,6 +68,12 @@ class CaffeLMDBRecordReader : public Reader {
 
     CaffeLMDBRecordReader();
 
+    //! return feed_data: not implemented
+    void feed_file_names(const std::vector<std::string>& file_names, size_t num_images, bool eos = false) override { return; }
+
+    //! return feed_raw_data: not implemented
+    void feed_data(const std::vector<unsigned char*>& images, const std::vector<size_t>& image_size, ExternalFileMode mode, bool eos = false, const std::vector<unsigned> roi_width = {}, const std::vector<unsigned> roi_height = {}, int width = 0, int height = 0, int channels = 0) override { return; }
```
Comment (shobiloganat) [inline rocAL/include/readers/image/caffe_lmdb_record_reader.h]: Please check the comments.

File: rocAL/include/readers/image/external_source_reader.h [hunk 20]
```diff
@@ -0,0 +1,120 @@
...
+    /// The loader will repeat images if necessary to be able to have images available in multiples of the load_batch_count,
+    /// for instance if there are 10 images in the dataset and _batch_count is 3, the loader repeats 2 images as if there are 12 images available.
+    size_t _batch_count = 1;
+    size_t _file_id = 0;
+    size_t _in_batch_read_count = 0;
+    bool _loop;
+    bool _shuffle;
+    int _read_counter = 0;
+    volatile bool _end_of_sequence;
+    //!< _file_count_all_shards total_number of files in to figure out the max_batch_size (usually needed for distributed training).
```
Comment (shobiloganat) [inline rocAL/include/readers/image/external_source_reader.h:107]: Why is this comment present here?

File: rocAL/source/api/rocal_api_data_loaders.cpp [hunk 21]
```diff
@@ -984,7 +923,7 @@ rocalJpegCOCOFileSource(
         output = context->master_graph->create_loader_output_tensor(info);
         auto cpu_num_threads = context->master_graph->calculate_cpu_num_threads(1);
 
-        context->master_graph->add_node<ImageLoaderNode>({}, {output})->init(internal_shard_count, cpu_num_threads, source_path, json_path, std::map<std::string, std::string>(), StorageType::COCO_FILE_SYSTEM, decType, shuffle, loop, context->user_batch_size(), context->master_graph->mem_type(), context->master_graph->meta_data_reader(), decoder_keep_original);
+        context->master_graph->add_node<ImageLoaderNode>({}, {output})->init(internal_shard_count, cpu_num_threads, source_path, json_path, std::map<std::string, std::string>(), StorageType::COCO_FILE_SYSTEM, decType, shuffle, loop, context->user_batch_size(), context->master_graph->mem_type(), context->master_graph->meta_data_reader(), decoder_keep_original, ExternalFileMode::FILENAME);
```
Comment (shobiloganat) [inline rocAL/source/api/rocal_api_data_loaders.cpp]: Keep externalfilemode as default argument and revert these changes.

File: rocAL/source/api/rocal_api_data_loaders.cpp [hunk 22]
```diff
@@ -2024,6 +1963,73 @@ rocalRawCIFAR10Source(
...
+        INFO("Internal buffer size width = " + TOSTR(width) + " height = " + TOSTR(height) + " depth = " + TOSTR(num_of_planes))
+        // Change the way ImageInfo is created for TensorInfo
+        // std::vector<size_t> dims = {context->user_batch_size(), height, width, num_of_planes};
+        auto info = TensorInfo(std::move(dims),
+                               context->master_graph->mem_type(),
+                               RocalTensorDataType::UINT8);
+        info.set_color_format(color_format);
+        info.set_tensor_layout(RocalTensorlayout::NHWC);
+        info.set_max_shape();
+        // auto info = ImageInfo(width, height,
```
Comment (shobiloganat) [inline rocAL/source/api/rocal_api_data_loaders.cpp]: remove the commented lines if not required.

File: rocAL/source/loaders/image/image_read_and_decode.cpp [hunk 23]
```diff
@@ -170,7 +191,52 @@ ImageReadAndDecode::load(unsigned char *buff,
             file_counter++;
         }
         //_file_load_time.end();// Debug timing
-        // return LoaderModuleStatus::OK;
```
Comment (shobiloganat) [inline rocAL/source/loaders/image/image_read_and_decode.cpp:173]: Check why this has been removed.

File: rocAL/source/meta_data/external_source_label_reader.cpp [hunk 24]
```diff
@@ -0,0 +1,94 @@
...
+    if (exists(image_name)) {
+        WRN("Entity with the same name exists")
+        return;
+    }
+    _map_content.insert(std::pair<std::string, std::shared_ptr<Label>>(image_name, info));
+}
+
+void ExternalSourceLabelReader::add_labels(std::vector<std::string> image_name, std::vector<int> label) {
+    if (image_name.size() != label.size()) {
+        std::cerr << "Image Name Size : " << image_name.size();
```
Comment (shobiloganat) [inline rocAL/source/meta_data/external_source_label_reader.cpp]: Remove std::cerr and use Throw

File: rocAL/source/readers/image/external_source_reader.cpp [hunk 25]
```diff
@@ -0,0 +1,234 @@
...
+            }
+            fseek(_current_fPtr, 0, SEEK_SET);  // Take the file pointer back to the start
+            _file_data[_curr_file_idx] = std::make_tuple((unsigned char*)next_file_name.data(), (size_t)_current_file_size, 0, 0, 0, 0, 0);
+            increment_read_ptr();
+        }
+    } else {
+        std::tuple<unsigned char*, size_t, int, int, int, unsigned, unsigned> image;
+        bool ret = pop_file_data(image);
+        if (_end_of_sequence && !ret) {
+            std::cerr << "\n EOS || POP FAILED ";
```
Comment (shobiloganat) [inline rocAL/source/readers/image/external_source_reader.cpp]: Change this to THROW

File: rocAL_pybind/amd/rocal/plugin/generic.py [hunk 26]
```diff
@@ -99,6 +99,37 @@ def __next__(self):
...
+                        "eos": self.eos}
+                    b.externalSourceFeedInput(*(kwargs_pybind.values()))
+                if self.loader._external_source_mode == types.EXTSOURCE_RAW_COMPRESSED:
+                    print(
```
Comment (shobiloganat) [inline rocAL_pybind/amd/rocal/plugin/generic.py]: Why mode 1 and mode 2 are not supported?

**Review by shobiloganat — COMMENTED** [review]
Please address review comments. 
Remove the usage of std::cerr with THROW.

File: rocAL/include/loaders/image/cifar10_data_loader.h [hunk 27]
```diff
@@ -47,6 +47,9 @@ class CIFAR10DataLoader : public LoaderModule {
     void shut_down() override;
     std::vector<std::vector<float>> &get_batch_random_bbox_crop_coords();
     void set_batch_random_bbox_crop_coords(std::vector<std::vector<float>> batch_crop_coords);
+    void feed_external_input(std::vector<std::string> input_images_names, bool labels, std::vector<unsigned char *> input_buffer,
```
Comment (fgladwin) [inline rocAL/include/loaders/image/cifar10_data_loader.h]: Add labels as optional bool

File: rocAL/include/loaders/image/image_loader.h [hunk 28]
```diff
@@ -87,4 +90,7 @@ class ImageLoader : public LoaderModule {
     bool _decoder_keep_original = false;
     int _device_id;
     size_t _max_tensor_width, _max_tensor_height;
+    size_t _max_decoded_width, _max_decoded_height;
```
Comment (fgladwin) [inline rocAL/include/loaders/image/image_loader.h]: Variables not used, to be removed

File: rocAL/include/loaders/image/image_read_and_decode.h (hunk 17 above)
Comment (fgladwin) [reply rocAL/include/loaders/image/image_read_and_decode.h]: @shobiloganat
Please let me know what changes to be done here

File: rocAL/include/loaders/image/node_image_loader.h [hunk 29]
```diff
@@ -40,7 +40,7 @@ class ImageLoaderNode : public Node {
     /// The loader will repeat images if necessary to be able to have images in multiples of the load_batch_count,
     /// for example if there are 10 images in the dataset and load_batch_count is 3, the loader repeats 2 images as if there are 12 images available.
     void init(unsigned internal_shard_count, unsigned cpu_num_threads, const std::string &source_path, const std::string &json_path, const std::map<std::string, std::string> feature_key_map, StorageType storage_type, DecoderType decoder_type, bool shuffle, bool loop,
-              size_t load_batch_count, RocalMemType mem_type, std::shared_ptr<MetaDataReader> meta_data_reader, bool decoder_keep_orig = false, const char *prefix = "", unsigned sequence_length = 0, unsigned step = 0, unsigned stride = 0);
+              size_t load_batch_count, RocalMemType mem_type, std::shared_ptr<MetaDataReader> meta_data_reader, bool decoder_keep_orig = false, ExternalFileMode external_file_mode = ExternalFileMode::FILENAME, const char *prefix = "", unsigned sequence_length = 0, unsigned step = 0, unsigned stride = 0);
```
Comment (fgladwin) [inline rocAL/include/loaders/image/node_image_loader.h]: Can the argument be added at the end?

File: rocAL/include/loaders/video/video_loader_sharded.h [hunk 30]
```diff
@@ -48,6 +48,9 @@ class VideoLoaderSharded : public LoaderModule {
     std::vector<size_t> get_sequence_start_frame_number() override;
     std::vector<std::vector<float>> get_sequence_frame_timestamps() override;
     Timing timing() override;
+    void feed_external_input(std::vector<std::string> input_images_names, bool labels, std::vector<unsigned char*> input_buffer,
+                             std::vector<unsigned> roi_width, std::vector<unsigned> roi_height,
+                             unsigned int max_width, unsigned int max_height, int channels, ExternalFileMode mode, bool eos) override { return; }
```
Comment (fgladwin) [inline rocAL/include/loaders/video/video_loader_sharded.h]: Remove return;

File: rocAL/include/loaders/video/video_loader.h [hunk 31]
```diff
@@ -56,6 +56,9 @@ class VideoLoader : public LoaderModule {
     std::vector<size_t> get_sequence_start_frame_number() override;
     std::vector<std::vector<float>> get_sequence_frame_timestamps() override;
     void shut_down() override;
+    void feed_external_input(std::vector<std::string> input_images_names, bool labels, std::vector<unsigned char*> input_buffer,
+                             std::vector<unsigned> roi_width, std::vector<unsigned> roi_height,
+                             unsigned int max_width, unsigned int max_height, int channels, ExternalFileMode mode, bool eos) override { return; }
```
Comment (fgladwin) [inline rocAL/include/loaders/video/video_loader.h]: Remove return

File: rocAL_pybind/amd/rocal/plugin/pytorch.py [hunk 32]
```diff
@@ -59,13 +59,74 @@ def __init__(self, pipeline, tensor_layout=types.NCHW, reverse_channels=False, m
...
+                        "roi_width": [],
+                        "roi_height": data_loader_source[2],
+                        "decoded_width": self.loader._external_source_user_given_width,
+                        "decoded_height": self.loader._external_source_user_given_height,
+                        "channels": 3,
+                        "external_source_mode": self.loader._external_source_mode,
+                        "rocal_tensor_layout": types.NCHW,
+                        "eos": self.eos}
+                    b.externalSourceFeedInput(*(kwargs_pybind.values()))
+                if self.loader._external_source_mode == types.EXTSOURCE_RAW_UNCOMPRESSED:
```
Comment (fgladwin) [inline rocAL_pybind/amd/rocal/plugin/pytorch.py]: Add else if

File: rocAL_pybind/rocal_pybind.cpp [hunk 33]
```diff
@@ -84,6 +84,32 @@ py::object wrapper_image_name(RocalContext context, int array_len) {
...
+    if (input_images_names.size() == 0) {  // Used for mode 1 and mode 2 for passing decoded buffers
+        size_t numArrays = py::len(arrays);
+        for (size_t i = 0; i < numArrays; i++) {
+            py::array_t<unsigned char> arr(arrays[i]);
+            py::buffer_info buf = arr.request();
+            uchar_arrays.push_back(static_cast<unsigned char *>(buf.ptr));
+        }
+    }
+
+    int status = rocalExternalSourceFeedInput(context, input_images_names, {1}, uchar_arrays, roi_width, roi_height, max_width, max_height, channels, mode, layout, eos);
```
Comment (fgladwin) [inline rocAL_pybind/rocal_pybind.cpp]: Remove the vector {1}

File: rocAL_pybind/rocal_pybind.cpp [hunk 34]
```diff
@@ -84,6 +84,32 @@ py::object wrapper_image_name(RocalContext context, int array_len) {
...
+    auto labels_tensor_list = rocalGetImageLabels(context);
```
Comment (fgladwin) [inline rocAL_pybind/rocal_pybind.cpp]: Add a comment

File: rocAL/include/readers/image/image_reader.h [hunk 35]
```diff
@@ -88,9 +96,10 @@ struct ReaderConfig {
     void set_file_prefix(const std::string &prefix) { _file_prefix = prefix; }
     std::string file_prefix() { return _file_prefix; }
     std::shared_ptr<MetaDataReader> meta_data_reader() { return _meta_data_reader; }
+    virtual ExternalFileMode mode() { return _file_mode; }
```
Comment (fgladwin) [inline rocAL/include/readers/image/image_reader.h]: Why is this virtual?

File: rocAL/include/readers/image/image_reader.h [hunk 36]
```diff
@@ -47,13 +47,20 @@ enum class StorageType {
...
+    RAWDATA_UNCOMPRESSED = 2,
 };
 
 struct ReaderConfig {
     explicit ReaderConfig(StorageType type, std::string path = "", std::string json_path = "",
                           const std::map<std::string, std::string> feature_key_map = std::map<std::string, std::string>(),
-                          bool shuffle = false, bool loop = false) : _type(type), _path(path), _json_path(json_path), _feature_key_map(feature_key_map), _shuffle(shuffle), _loop(loop) {}
-    virtual StorageType type() { return _type; };
+                          bool shuffle = false, bool loop = false, ExternalFileMode mode = ExternalFileMode::FILENAME) : _storage_type(type), _path(path), _json_path(json_path), _feature_key_map(feature_key_map), _shuffle(shuffle), _loop(loop), _file_mode(mode) {}
+    virtual StorageType storage_type() { return _storage_type; }
```
Comment (fgladwin) [inline rocAL/include/readers/image/image_reader.h]: Why is it virtual?

File: rocAL/include/readers/image/file_source_reader.h [hunk 37]
```diff
@@ -64,6 +64,12 @@ class FileSourceReader : public Reader {
 
     FileSourceReader();
 
+    //! return feed_data: not implemented
+    void feed_file_names(const std::vector<std::string> &file_names, size_t num_images, bool eos = false) override { return; }
+
+    //! return feed_raw_data: not implemented
+    void feed_data(const std::vector<unsigned char *> &images, const std::vector<size_t> &image_size, ExternalFileMode mode, bool eos = false, const std::vector<unsigned> roi_width = {}, const std::vector<unsigned> roi_height = {}, int width = 0, int height = 0, int channels = 0) override { return; }
```
Comment (fgladwin) [inline rocAL/include/readers/image/file_source_reader.h]: Throw error saying not implemented

File: rocAL/source/loaders/image/image_read_and_decode.cpp [hunk 38]
```diff
@@ -142,7 +162,8 @@ ImageReadAndDecode::load(unsigned char *buff,
     const unsigned output_planes = std::get<1>(ret);
     const bool keep_original = decoder_keep_original;
     const size_t image_size = max_decoded_width * max_decoded_height * output_planes * sizeof(unsigned char);
-
+    bool is_external_source = (_storage_type == StorageType::EXTERNAL_FILE_SOURCE);
```
Comment (fgladwin) [inline rocAL/source/loaders/image/image_read_and_decode.cpp]: Can this be a private variable? Instead of checking each time check once in init

File: rocAL/source/loaders/image/image_read_and_decode.cpp (hunk 23 above)
Comment (fgladwin) [reply rocAL/source/loaders/image/image_read_and_decode.cpp:173]: This is a commented statement

File: rocAL/source/meta_data/external_source_label_reader.cpp (hunk 24 above)
Comment (fgladwin) [reply rocAL/source/meta_data/external_source_label_reader.cpp]: This is not used hence removed

File: rocAL/source/readers/image/external_source_reader.cpp (hunk 25 above)
Comment (fgladwin) [reply rocAL/source/readers/image/external_source_reader.cpp]: Changed to warning

File: rocAL_pybind/amd/rocal/plugin/generic.py (hunk 26 above)
Comment (fgladwin) [reply rocAL_pybind/amd/rocal/plugin/generic.py]: In the generic iterator it is not being supported I am not sure
@swetha097 Please comment

File: rocAL/include/loaders/image/cifar10_data_loader.h (hunk 27 above)
Comment (fgladwin) [reply rocAL/include/loaders/image/cifar10_data_loader.h]: Labels bool need not be passed to loaders, hence removing it

File: rocAL/include/loaders/image/node_image_loader.h (hunk 29 above)
Comment (fgladwin) [reply rocAL/include/loaders/image/node_image_loader.h]: This is the correct position which will lead to minimal changes - Fiona

File: rocAL/include/readers/image/image_reader.h (hunk 35 above)
Comment (fgladwin) [reply rocAL/include/readers/image/image_reader.h]: Removed

File: rocAL/source/readers/image/external_source_reader.cpp [hunk 39]
```diff
@@ -0,0 +1,234 @@
...
+    _shuffle = false;
+    _file_count_all_shards = 0;
+    _loop = false;  // loop not supported for external source
+    _end_of_sequence = false;
+}
+
+// return batch_size() for count_items unless end_of_sequence has been signalled.
+unsigned ExternalSourceReader::count_items() {
+    if (_file_mode == ExternalFileMode::FILENAME) {
+        if (_end_of_sequence && _file_names_queue.empty()) {
```
Comment (fgladwin) [inline rocAL/source/readers/image/external_source_reader.cpp:43]: Should this be OR or AND?

File: rocAL/source/readers/image/external_source_reader.cpp [hunk 40]
```diff
@@ -0,0 +1,234 @@
...
+    _folder_path = desc.path();
+    _file_id = 0;
+    _shard_id = desc.get_shard_id();
+    _shard_count = desc.get_shard_count();
+    _batch_count = desc.get_batch_size();
+    _shuffle = desc.shuffle();
+    _loop = desc.loop();
+    _file_mode = desc.mode();
+    _end_of_sequence = false;
+    _file_data.reserve(_batch_count);
```
Comment (fgladwin) [inline rocAL/source/readers/image/external_source_reader.cpp:65]: Sharding support not given

File: rocAL/source/loaders/image/image_read_and_decode.cpp (hunk 38 above)
Comment (fgladwin) [reply rocAL/source/loaders/image/image_read_and_decode.cpp]: Done

File: rocAL_pybind/amd/rocal/plugin/generic.py (hunk 15 above)
Comment (swetha097) [reply rocAL_pybind/amd/rocal/plugin/generic.py]: I think this has to be changed, to be passed from the python end
